---
title: "Excercise 4"
author: "Yaal Shapira"
date: "18/11/2021"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(compGenomRData)
library(DESeq2)
library(gProfileR)
library(gage)
library(ggplot2)
library(wesanderson)
library(stringr)
library(topGO)
library(biomaRt)
library(org.Hs.eg.db)

```

Loading the data:
```{r loading, warning=FALSE}
counts_file <- system.file("extdata/rna-seq/SRP029880.raw_counts.tsv",
                           package = "compGenomRData")
coldata_file <- system.file("extdata/rna-seq/SRP029880.colData.tsv",
                            package = "compGenomRData")
counts <- as.matrix(read.table(counts_file, header = T, sep = '\t'))

#remove the 'width' column
countData <- as.matrix(subset(counts, select = c(-width)))

#define the experimental setup
colData <- read.table(coldata_file, header = T, sep = '\t',
                      stringsAsFactors = TRUE)

#define the design formula
designFormula <- "~ group"

#create a DESeq dataset object from the count matrix and the colData
dds <- DESeqDataSetFromMatrix(countData = countData,
                              colData = colData,
                              design = as.formula(designFormula))
dds <- DESeq(dds)

DEresults = results(dds, contrast = c("group", 'CASE', 'CTRL'))
```

Let’s select the genes that are significantly differentially expressed between the case and control samples. Let’s extract genes that have an adjusted p-value below 0.1 and that show a 2-fold change (either negative or positive) in the case compared to control. We will then feed this gene set into the gProfileR function.

```{r selection, warning=FALSE}
#remove genes with NA values
DE <- DEresults[!is.na(DEresults$padj),]

genestested = setNames( DE@listData[["pvalue"]],rownames(DE))
#select genes with adjusted p-values below 0.1
DE <- DE[DE$padj < 0.1,]

#select genes with absolute log2 fold change above 1 (two-fold change)
DE <- DE[abs(DE$log2FoldChange) > 1,]

#get the list of genes of interest
genesOfInterest <- rownames(DE)

#calculate enriched GO terms
goResults <- gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'GO',
                       hier_filtering = 'moderate')
```

1. Re-run gProfileR, this time using pathway annotations such as KEGG, REACTOME, and protein complex databases such as CORUM, in addition to the GO terms. Sort the resulting tables by columns `precision` and/or `recall`. How do the top GO terms change when sorted for `precision`, `recall`, or `p.value`? hint: use `order()` for sorting. [Difficulty: **Beginner**]

```{r 1, warning = F}
KeggResults = gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'KEGG',
                       hier_filtering = 'moderate')
  
ReactomeResults = gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'REAC',
                       hier_filtering = 'moderate')
  
CorumRersults = gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'CORUM',
                       hier_filtering = 'moderate')

goResults_pre = goResults[order(goResults$precision),"term.name"]
goResults_rec = goResults[order(goResults$recall),"term.name"]
goResults_pval = goResults[order(goResults$p.value),"term.name"]

go_topterms = data.frame(Precision = goResults_pre, P.Value = goResults_pval, Recall = goResults_rec)

knitr::kable(head(go_topterms), format = "markdown")
```
#### Gene set enrichment analysis

We use the bioconductor package gage to demonstrate how to do GSEA using normalized expression data of the samples as input.

```{r}
#Let's define the first gene set as the list of genes from one of the
#significant GO terms found in the GO analysis. order go results by pvalue
goResults <- goResults[order(goResults$p.value),]

#restrict the terms that have at most 100 genes overlapping with the query
go <- goResults[goResults$overlap.size < 100,]

# use the top term from this table to create a gene set
geneSet1 <- unlist(strsplit(go[1,]$intersection, ','))

#Define another gene set by just randomly selecting 25 genes from the counts
#table get normalized counts from DESeq2 results
normalizedCounts <- DESeq2::counts(dds, normalized = TRUE)

geneSet2 <- sample(rownames(normalizedCounts), 25)

geneSets <- list('top_GO_term' = geneSet1,
                 'random_set' = geneSet2)

# Using the defined gene sets, we’d like to do a group comparison between the case
# samples with respect to the control samples.

#Use the normalized counts to carry out a GSEA.
gseaResults <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'as.group')
```

2. Repeat the gene set enrichment analysis by trying different options for the `compare` argument of the `GAGE:gage`
function. How do the results differ? [Difficulty: **Beginner**]
```{r 2}
gseaResults_paired <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'paired')
gseaResults_unpaired <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'unpaired')
gseaResults_1ongroup <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = '1ongroup')
```
 The results are the same however we compare the groups. the difference is in the way of the comparison between the groups. e.g. the "as.group" compares the average between the groups, while "1ongroup" compares the value of each sample against the reference average.


3. Make a scatter plot of GO term sizes and obtained p-values by setting the `gProfiler::gprofiler` argument `significant = FALSE`. Is there a correlation of term sizes and p-values? (Hint: Take -log10 of p-values). If so, how can this bias be mitigated? [Difficulty: **Intermediate**]

```{r 3, warning = F}
plot_goResults = gprofiler(query = genesOfInterest,
                       organism = 'hsapiens',
                       src_filter = 'GO',
                       hier_filtering = 'moderate',
                       significant = F)
plot_goResults$neglog.pval = -log10(plot_goResults$p.value)
ggplot(data = plot_goResults, aes(x = plot_goResults[,"term.size"],
                                  y = plot_goResults[,"neglog.pval"],
                                  color = str_wrap(plot_goResults[,"term.name"],50))) +
  geom_point(size = 2) + labs(title = "term size vs negative log p.value", color = "function") + xlab("Term size") + ylab("-log10(p.value)")
```
We see a strong correlation between term size to small p.value. This can be mitigated with some kind of adjusted p.value, as basically for each term there are multiple testing and of course the more testing we will find more genes that are enriched in this term.





4. Do a gene-set enrichment analysis using gene sets from top 10 GO terms. [Difficulty: **Intermediate**]
```{r 4}
top_10_GO_terms = unlist(strsplit(goResults[order(goResults$p.value),"intersection"][1:10], ','))

geneSetsII = list('top_GO_term' = top_10_GO_terms,
                 'random_set' = geneSet2)

gseaResults_top10 <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = geneSets, compare = 'as.group')
gseaResults_top10
```


5. What are the other available R packages that can carry out gene set enrichment analysis for RNA-seq datasets? [Difficulty: **Intermediate**]
MAVTgsa, ClusterProfiler, GSEABase, RGSEA



6.  Use the topGO package (https://bioconductor.org/packages/release/bioc/html/topGO.html) to re-do the GO term analysis. Compare and contrast the results with what has been obtained using the `gProfileR` package. Which tool is faster, `gProfileR` or topGO? Why? [Difficulty: **Advanced**]
```{r 6}
geneSelFunc <- function (score) {
    return(score <= 0.01)
}

topgo_data = new("topGOdata",
                 ontology = "BP",
                 allGenes = genestested,
                 geneSel = geneSelFunc,
                 annot = annFUN.org, 
                 mapping = "org.Hs.eg", 
                 ID = "symbol"
                 )
resultfishr = runTest(topgo_data, algorithm = "classic", statistic = "fisher")
topgo_res = GenTable(topgo_data, ClassicFisher = resultfishr)
topgo = topgo_res[,"Term"]
topgo_vs_gprofiler = data.frame(gProfiler = goResults_pval[1:10], TopGo = topgo)

knitr::kable(topgo_vs_gprofiler, format = "markdown")

```

7. Given a gene set annotated for human, how can it be utilized to work on _C. elegans_ data? (Hint: See `biomaRt::getLDS`). [Difficulty: **Advanced**]
you could, using getLDS, link human gene database and C.elegans gene database. Then, transform the c.elegans DEGs into human homologs and run a Gene enrichment analysis.

8. Import curated pathway gene sets with Entrez identifiers from the [MSIGDB database](http://software.broadinstitute.org/gsea/msigdb/collections.jsp) and re-do the GSEA for all curated gene sets. [Difficulty: **Advanced**]
```{r 8 }
#read table
H_set  = read.table("C:/Users/domin/Desktop/R/GenomeW4/h.all.v7.4.entrez.gmt", fill = T)
hs = org.Hs.eg.db
#create a list with SYMBOL instead of entrez
names(H_set) = NULL
H_set = (H_set[3:ncol(H_set)])

H_vector = as.vector(t(H_set))
H_vector = as.character(H_vector[!is.na(H_vector)])

H_symbol = select(hs,
                    keys = H_vector,
                    columns = "SYMBOL",
                    keytype = "ENTREZID")["SYMBOL"]

H_symbol = as.vector(t(H_symbol[!is.na(H_symbol)]))



#create the geneset, remove genes that weren't measured

H_geneset = intersect(H_symbol, rownames(normalizedCounts))
  
H_geneSets <- list('H gene set' = H_geneset[sample(length(H_geneset),25)],
                 'random_set' = geneSet2)


Hset_GSEA <- gage(exprs = log2(normalizedCounts+1),
                    ref = match(rownames(colData[colData$group =='CTRL',]),
                                colnames(normalizedCounts)),
                    samp = match(rownames(colData[colData$group == 'CASE',]),
                                 colnames(normalizedCounts)),
                    gsets = H_geneSets, compare = 'as.group')

```




